zcash_primitives/src/consensus.rs

```diff
@@ -484,6 +497,70 @@ impl BranchId {
         // Sprout rules apply before any network upgrade
         BranchId::Sprout
     }
+
+    pub fn min_height<P: Parameters>(&self, params: &P) -> Option<BlockHeight> {
+        match self {
+            BranchId::Sprout => Some(BlockHeight(1)),
+            BranchId::Overwinter => params.activation_height(NetworkUpgrade::Overwinter),
+            BranchId::Sapling => params.activation_height(NetworkUpgrade::Sapling),
+            BranchId::Blossom => params.activation_height(NetworkUpgrade::Blossom),
+            BranchId::Heartwood => params.activation_height(NetworkUpgrade::Heartwood),
+            BranchId::Canopy => params.activation_height(NetworkUpgrade::Canopy),
+            BranchId::Nu5 => params.activation_height(NetworkUpgrade::Nu5),
+            #[cfg(feature = "zfuture")]
+            BranchId::ZFuture => params.activation_height(NetworkUpgrade::ZFuture),
+        }
+    }
+
+    pub fn max_height<P: Parameters>(&self, params: &P) -> Option<BlockHeight> {
+        (match self {
+            BranchId::Sprout => params.activation_height(NetworkUpgrade::Overwinter),
+            BranchId::Overwinter => params.activation_height(NetworkUpgrade::Sapling),
+            BranchId::Sapling => params.activation_height(NetworkUpgrade::Blossom),
+            BranchId::Blossom => params.activation_height(NetworkUpgrade::Heartwood),
+            BranchId::Heartwood => params.activation_height(NetworkUpgrade::Canopy),
+            BranchId::Canopy => params.activation_height(NetworkUpgrade::Nu5),
+            BranchId::Nu5 => None,
+            #[cfg(feature = "zfuture")]
+            BranchId::ZFuture => None,
```

Comment:
should these heights all be 1 less?

Reply:
... they are! https://github.com/zcash/librustzcash/pull/385/files#diff-390655b2b7dbbfda727fa47b40e98c43fd479427f293da891c00b708ba01c9b7R527